Sources/App/Core/PackageCollection+generate.swift

```diff
@@ -15,42 +15,49 @@ extension PackageCollection {
 
-    static func generate(db: Database,
-                         packageURLs: [String],
-                         authorName: String? = nil,
-                         collectionName: String,
-                         keywords: [String]? = nil,
-                         overview: String? = nil,
-                         revision: Int? = nil) -> EventLoopFuture<PackageCollection> {
-        packageQuery(db: db)
-            .filter(\.$url ~~ packageURLs)
-            .all()
-            .mapEachCompact { Package.init(package:$0, keywords: keywords) }
-            .map {
-                PackageCollection.init(
-                    name: collectionName,
-                    overview: overview,
-                    keywords: keywords,
-                    packages: $0,
-                    formatVersion: .v1_0,
-                    revision: revision,
-                    generatedAt: Current.date(),
-                    generatedBy: authorName.map(Author.init(name:)))
-            }
+    enum Filter {
+        case urls([String])
+        case author(String)
     }
 
     static func generate(db: Database,
-                         owner: String,
+                         filterBy filter: Filter,
                          authorName: String? = nil,
                          collectionName: String,
                          keywords: [String]? = nil,
                          overview: String? = nil,
                          revision: Int? = nil) -> EventLoopFuture<PackageCollection> {
-        packageQuery(db: db)
+        var query = App.Version.query(on: db)
+            .with(\.$builds)
+            .with(\.$products)
+            .with(\.$targets)
+            .with(\.$package) {
+                $0.with(\.$repositories)
+            }
+            .join(App.Package.self, on: \App.Package.$id == \Version.$package.$id)
             .join(Repository.self, on: \App.Package.$id == \Repository.$package.$id)
-            .filter(Repository.self, \.$owner, .custom("ilike"), owner)
-            .all()
-            .mapEachCompact { Package.init(package:$0, keywords: keywords) }
+            .filter(Version.self, \.$latest ~~ [.release, .preRelease])
+
+        switch filter {
+            case let .author(owner):
+                query = query
+                    .filter(Repository.self, \.$owner, .custom("ilike"), owner)
+            case let .urls(packageURLs):
+                query = query
+                    .filter(App.Package.self, \.$url ~~ packageURLs)
+        }
+
+        return query.all()
+            .map { versions in
+                // Multiple versions can reference the same package, therefore
+                // we need to group them so we don't create duplicate packages.
+                // This requires App.Package to be Hashable and Equatable.
+                Dictionary(grouping: versions, by: { $0.package })
+                    .sorted(by: { $0.key.url < $1.key.url })
```

Comment:
I tried to avoid making `Package` conform to `Equatable` and `Hashable`, because it’s quite a sweeping change and basing it solely on `id` comparison (while valid in this case) may not be ideal generally.

I used an `IDHashable<T: Identifiable>` container and while it worked it caused some (solvable) downstream issues. (For instance, `Package` would need to be `Identifiable` which, while seemingly straightforward doesn’t compile.)

A non-generic variant worked but I’m not sure it’s an improvement over just making `Package` conform directly.